runtime/gc_base/ObjectAccessBarrier.cpp

```diff
@@ -1513,25 +1514,26 @@ MM_ObjectAccessBarrier::compareAndSwapObject(J9VMThread *vmThread, J9Object *des
 		actualDestAddress = J9OAB_MIXEDOBJECT_EA(destObject, ((UDATA)destAddress - (UDATA)destObject), fj9object_t);
 	}
 
-	/* Note: This is a bit of a special case -- we call preObjectStore even though the store
-	 * may not actually occur. This is safe and correct for Metronome.
-	 */
-	preObjectStore(vmThread, destObject, actualDestAddress, swapObject, true);
-	protectIfVolatileBefore(vmThread, true, false, false);
-	
-	/* One of these conditions will evaluate to true at compile time based on whether fields are 32 or 64 bits */
-	bool result = false;
-	if (sizeof(fj9object_t) == sizeof(UDATA)) {
-		result = ((UDATA)compareValue == MM_AtomicOperations::lockCompareExchange((UDATA *)actualDestAddress, (UDATA)compareValue, (UDATA)swapValue));
-	} else if (sizeof(fj9object_t) == sizeof(U_32)) {
-		result = ((U_32)(UDATA)compareValue == MM_AtomicOperations::lockCompareExchangeU32((U_32 *)actualDestAddress, (U_32)(UDATA)compareValue, (U_32)(UDATA)swapValue));
-	} else {
-		assume(false, "fj9object_t must be either U_32 (compressed pointers / 32-bit VM) or UDATA");
-	}
-	
-	protectIfVolatileAfter(vmThread, true, false, false);
-	if (result) {
-		postObjectStore(vmThread, destObject, actualDestAddress, swapObject, true);
+	if (preObjectRead(vmThread, destObject, actualDestAddress)) {
+		/* Note: This is a bit of a special case -- we call preObjectStore even though the store
+		 * may not actually occur. This is safe and correct for Metronome.
+		 */
+		preObjectStore(vmThread, destObject, actualDestAddress, swapObject, true);
+		protectIfVolatileBefore(vmThread, true, false, false);
+
+		/* One of these conditions will evaluate to true at compile time based on whether fields are 32 or 64 bits */
+		if (sizeof(fj9object_t) == sizeof(UDATA)) {
```

Comment:
Could this just become a compile time check with a #error in the else?

```
#if (sizeof(fj9object) == sizeof(UDATA))
....
```

Not sure if sizeof gets evaluated at the right time though.

Reply:
I did not change this (it only got indented), but if we are to change, I believe it the cleanest is to relay on 
#if defined (OMR_GC_COMPRESSED_POINTERS)
and then there is no need to error clause at all

Reply:
Replaced the check with J9VM_GC_COMPRESSED_POINTERS (and swapped the order of two lines)